src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBase.cs

```diff
@@ -501,33 +584,141 @@ internal override void OnItemClicked(int clickedIndex)
 						// The CollectionView may have intercepted the change
 						clickedIndex = collectionView.CurrentPosition;
 					}
-					SelectedIndex = clickedIndex;
+
+					if (modifiers.HasFlag(VirtualKeyModifiers.Control) && clickedIndex == SelectedIndex)
+					{
+						SelectedIndex = -1;
+					}
+					else
+					{
+						SelectedIndex = clickedIndex;
+					}
+
 					break;
 				case ListViewSelectionMode.Multiple:
+					if (!modifiers.HasFlag(VirtualKeyModifiers.Shift))
+					{
+						FlipSelectionInMultipleSelection(clickedIndex, item: clickedItem, container: clickedContainer);
+						break;
+					}
+
+					// if shift is held, all items between the clicked item and the previously
+					// focused item match the selection of the previously focused item
+					var newSelection = focusedContainer?.IsSelected ?? false;
+
+					var multipleLowerBound = Math.Min(focusedIndex, clickedIndex);
+					var multipleUpperBound = Math.Max(focusedIndex, clickedIndex);
+
+					for (int currentIndex = multipleLowerBound; currentIndex <= multipleUpperBound; currentIndex++)
+					{
+						if (ContainerFromIndex(currentIndex) is SelectorItem container && container.IsSelected != newSelection)
+						{
+							FlipSelectionInMultipleSelection(currentIndex, container: container);
+						}
+					}
+
+					break;
 				case ListViewSelectionMode.Extended:
```

Comment:
these blocks are getting rather long, can you extract them into methods? (they can be local method)